### PR TITLE
03 data structure

### PR DESCRIPTION
**Pull request by Romservik29**

### CODE REVIEW

File: homeworks/01-bubble-sort/src/App.tsx [hunk 1]
```diff
@@ -24,42 +24,42 @@ export class App extends Component<{}, State> {
...
+          this.setState({ interval: null, sorting: Sorting.Finished })
+        }
+      }, speed)
+      this.setState({ interval: interval, sorting: Sorting.Running, speed: speed })
     }
-    else this.setState({speed: speed})
+    else this.setState({ speed: speed })
   }
   render() {
     return (
```
Comment (Romservik29) [inline homeworks/01-bubble-sort/src/App.tsx:65]: Упс!

File: homeworks/01-bubble-sort/src/App.tsx [hunk 2]
```diff
@@ -24,42 +24,42 @@ export class App extends Component<{}, State> {
   constructor(props: {}) {
     super(props)
     this.state = {
-      arr: { arr: [...rndArray(30, 200)]},
```
Comment (shpakkdv) [inline homeworks/01-bubble-sort/src/App.tsx:27]: All the changes in this file should be done in a separate Pull Request, or separate branch, or just a separate commit in your main branch.

File: homeworks/03-data-structure/algorithms/binary-search-tree.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+interface BinarySearch extends IBinaryTree<number> {
+    has: (searchVal: number, tree: TreeNode<number>) => boolean
+}
+
+export class BinarySearchTree extends BinaryTree<number> implements BinarySearch {
+    constructor(rootNode: TreeNode<number>) {
+        super(rootNode)
+    }
+    has(serachValue: number, tree: TreeNode<number> = this.rootTree): boolean {
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-search-tree.ts:11]: Always define access modifiers.

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+  breadth: (node: TreeNode<T>) => T[]
+}
+
+interface TreeNode<T> {
+  value: T;
+  left: TreeNode<T> | null;
+  right: TreeNode<T> | null;
+}
+interface IBinaryTree<T> {
+  setTree(tree: TreeNode<T>): this;
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:24]: `constructor` is absent.

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    switch (traverse) {
+      case Traverse.Inorder:
+        return this.inorder(tree);
+      case Traverse.Postorder:
+        return this.postorder(tree);
+      case Traverse.Preorder:
+        return this.preorder(tree);
+      case Traverse.Breadth:
+        return this.breadth(tree);
+      default: return assertNever(traverse);
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:51]: Nice to see 👍

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  right: TreeNode<T> | null;
+}
+interface IBinaryTree<T> {
+  setTree(tree: TreeNode<T>): this;
+  traverse(travesrse: Traverse): T[];
+  getColumn(columnOrder: number): T[];
+}
+
+class BinaryTree<T> implements IBinaryTree<T>, Dfs<T> {
+  rootTree: TreeNode<T>
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:30]: Always define access modifiers.

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  Preorder,
+  Inorder,
+  Postorder,
+  Breadth
+}
+const assertNever = (arg: never): never => {
+  throw new Error(`${arg}`);
+};
+
+interface Dfs<T> {
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:11]: Better to store interfaces in a separate file.

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 8]
```diff
@@ -0,0 +1,111 @@
+enum Traverse {
+  Preorder,
+  Inorder,
+  Postorder,
+  Breadth
+}
+const assertNever = (arg: never): never => {
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:7]: Better to store utils in a separate file.

File: homeworks/03-data-structure/algorithms/binary-tree.ts [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+  left: TreeNode<T> | null;
+  right: TreeNode<T> | null;
+}
+interface IBinaryTree<T> {
+  setTree(tree: TreeNode<T>): this;
+  traverse(travesrse: Traverse): T[];
+  getColumn(columnOrder: number): T[];
+}
+
+class BinaryTree<T> implements IBinaryTree<T>, Dfs<T> {
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/binary-tree.ts:29]: You don't need `Dfs<T>`. All those methods should be private methods.

File: homeworks/03-data-structure/algorithms/merge-sort.ts [hunk 10]
```diff
@@ -0,0 +1,24 @@
+type CompareFunction<T> = (val1: T, val2: T) => 1 | 0 | -1
```
Comment (shpakkdv) [inline homeworks/03-data-structure/algorithms/merge-sort.ts:1]: It should return any number.

File: homeworks/03-data-structure/tests/binary-search-tree.test.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        right: {
+            value: 7,
+            left: null,
+            right: null
+        },
+    },
+}
+const searchTree = new BinarySearchTree(tree)
+
+it("has",()=>{
```
Comment (shpakkdv) [inline homeworks/03-data-structure/tests/binary-search-tree.test.ts:37]: Should be `describe` before.

File: homeworks/03-data-structure/tests/binary-search-tree.test.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+        expect(searchTree.has(8)).toBe(false)
+        expect(searchTree.has(5)).toBe(true)
+        expect(searchTree.has(5,{} as TreeNode<number>)).toBe(false)
```
Comment (shpakkdv) [inline homeworks/03-data-structure/tests/binary-search-tree.test.ts:40]: Tests should be separate and independent.

File: homeworks/03-data-structure/tests/binary-tree.test.ts [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+            right: null
+        },
+    },
+}
+//               -2-1   0   1 2 
+//                      1
+//                  2       3
+//                4   5   6   7
+let binaryTree = new BinaryTree(tree)
+it("create binary tree", () => {
```
Comment (shpakkdv) [inline homeworks/03-data-structure/tests/binary-tree.test.ts:39]: All `it`s should be under `describe`.

File: homeworks/03-data-structure/tests/binary-tree.test.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+it('bfs', () => {
+    expect(binaryTree.breadth(binaryTree.rootTree)).toEqual([1, 2, 3, 4, 5, 6, 7])
+})
+
+it("traverse", () => {
+    expect(binaryTree.traverse(Traverse.Preorder, binaryTree.rootTree)).toStrictEqual([1, 2, 4, 5, 3, 6, 7])
+    expect(binaryTree.traverse(Traverse.Inorder, binaryTree.rootTree)).toStrictEqual([2, 4, 5, 1, 3, 6, 7])
+    expect(binaryTree.traverse(Traverse.Postorder, binaryTree.rootTree)).toStrictEqual([4, 5, 2, 6, 7, 3, 1])
+    expect(binaryTree.traverse(Traverse.Breadth)).toEqual([1, 2, 3, 4, 5, 6, 7])
+    expect(() => binaryTree.traverse({} as never)).toThrowError();
```
Comment (shpakkdv) [inline homeworks/03-data-structure/tests/binary-tree.test.ts:64]: Tests should be separate and independent. Every test should check only one functionality and has an understandable description.

**Review by shpakkdv — APPROVED** [review]

File: homeworks/03-data-structure/algorithms/binary-search-tree.ts (hunk 3 above)
Comment (Romservik29) [reply homeworks/03-data-structure/algorithms/binary-search-tree.ts:11]: Should I define obviously public for better readable?
